guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,158 @@
+
+Hybrid concurrency limits the damage from a segmentation fault or other error that kills a process. A single process dying will kill that process's threads but not threads in other processes. Each process has its own independent garbage collector.
+
+Choosing Default Settings
+-------------------------
+
+Rails' default settings are chosen for small applications. You can improve performance for your large application that serves a lot of requests by changing settings.
+
+This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.
+
+These are production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
+
+[Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) may also be useful.
+
+### Threads Per Process
+
+Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Discourse also executes many database queries per request and frequently uses Redis. More self-contained applications with fewer database and API queries benefit from around 3 threads per process.
+
+The default Puma configuration mentions that "as a rule of thumb, increasing the number of threads will increase how much traffic a given process can handle (throughput), but due to CRuby's Global VM Lock (GVL) it has diminishing returns and will degrade the response time (latency) of the application."
+
+To set the number of threads, you can change the call to the +threads+ method in +config/puma.rb+. Or you can set the +RAILS_MAX_THREADS+ environment variable, which will do the same.
+
+### Number of Processes
+
+When using hybrid threads and processes, you should run 1 process per available processor core. Automatic methods to determine the number of cores are unreliable. You should specify the number of processes manually.
+
+To set the number of worker processes, you can change the call to the +workers+ method in +config/puma.rb+. Or you can set the +WEB_CONCURRENCY+ environment variable, which will do the same.
+
+### Preloading
+
+Puma creates new workers from a master process. By loading your application code in the master process, you can avoid doing so after creating the worker. This permits sharing more memory across processes.
+
+In a few cases it may not make sense to preload your application. In that case it's possible to turn off application preloading.
+
+Puma preloads your application by default by calling +preload_app!+ in +config/puma.rb+. If you remove this call, your application will not be preloaded.
```

Comment:
This might not always be true, since in some cases [Puma preload the app by default](https://github.com/puma/puma/blob/c2aa7371c64fe1867574f67b865ce42f01a596b5/README.md?plain=1#L125-L137)

Reply:
Ah, good catch. I'll mention that it preloads if the CLI arg is specified, too.